### PR TITLE
[AMDGPU] Fix selection of s_load_b96 on GFX11

### PR DESCRIPTION
**Pull request by jayfoad**
Fix a bug which resulted in selection of s_load_b96 on GFX11, which only
exists in GFX12.

The root cause was a mismatch between legalization and selection. The
condition used to check that the load was uniform in legalization
(SITargetLowering::LowerLOAD) was "!Op->isDivergent()". The condition
used to detect a non-uniform load during selection
(AMDGPUDAGToDAGISel::isUniformLoad()) was
"N->isDivergent() && !AMDGPUInstrInfo::isUniformMMO(MMO)". This makes a
difference when IR uniformity analysis has more information than SDAG's
built in analysis. In the test case this is because IR UA reports that
everything is uniform if isSingleLaneExecution() returns true, e.g. if
the specified max flat workgroup size is 1, but SDAG does not have this
optimization.

The immediate fix is to use the same condition to detect uniform loads
in legalization and selection. In future SDAG should learn about
isSingleLaneExecution(), and then it could probably stop relying on IR
metadata to detect uniform loads.

### CODE REVIEW

**Comment by jayfoad** [comment]
> In future SDAG should learn about isSingleLaneExecution(),

I've raised #108038 for this.

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -10334,24 +10335,24 @@ SDValue SITargetLowering::LowerLOAD(SDValue Op, SelectionDAG &DAG) const {
...
+      // have the same legalization requirements as global and private
+      // loads.
+      //
     }
-    // Non-uniform loads will be selected to MUBUF instructions, so they
-    // have the same legalization requirements as global and private
-    // loads.
-    //
   }
+
```
Comment (piotrAMD) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:10355]: Nit: Spurious extra line.

**Review by piotrAMD — COMMENTED** [review]
LGTM. 
I assume global-isel doesn't have this problem?

**Review by piotrAMD — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp (hunk 1 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:10355]: I did that on purpose to separate these two cases (which are guarded by `(!Op->isDivergent() || AMDGPUInstrInfo::isUniformMMO(MMO)) && ...`) from the next case.

**Comment by jayfoad** [comment]
> LGTM. I assume global-isel doesn't have this problem?

I think GISel is OK because it is already calling `AMDGPUInstrInfo::isUniformMMO` from `AMDGPURegisterBankInfo::isScalarLoadLegal`.

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp (hunk 1 above)
Comment (piotrAMD) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:10355]: Ok, makes sense.

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -10334,24 +10335,24 @@ SDValue SITargetLowering::LowerLOAD(SDValue Op, SelectionDAG &DAG) const {
...
-      AS == AMDGPUAS::CONSTANT_ADDRESS_32BIT ||
-      AS == AMDGPUAS::GLOBAL_ADDRESS) {
-    if (Subtarget->getScalarizeGlobalBehavior() && !Op->isDivergent() &&
-        Load->isSimple() && isMemOpHasNoClobberedMemOperand(Load) &&
-        Alignment >= Align(4) && NumElements < 32) {
-      if (MemVT.isPow2VectorType() ||
-          (Subtarget->hasScalarDwordx3Loads() && NumElements == 3))
-        return SDValue();
-      return WidenOrSplitVectorLoad(Op, DAG);
+    if (AS == AMDGPUAS::CONSTANT_ADDRESS ||
```
Comment (perlfu) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:10339]: `AMDGPUAS::CONSTANT_ADDRESS` and `AMDGPUAS::CONSTANT_ADDRESS_32BIT` do nothing here as they are completely handled by the previous `if()`?
If so, then you can hoist inner `if(Subtarget...` and merge `AS == ...` condition into it?

File: llvm/test/CodeGen/AMDGPU/load-constant-always-uniform.ll [hunk 3]
```diff
@@ -0,0 +1,42 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1100 < %s | FileCheck %s -prefix=GFX11
```
Comment (perlfu) [inline llvm/test/CodeGen/AMDGPU/load-constant-always-uniform.ll:2]: Perhaps this should also have `gfx1200` run line to show that selection on GFX12 uses `s_load_b96`?

File: llvm/test/CodeGen/AMDGPU/load-constant-always-uniform.ll [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  %i5 = getelementptr <{ [4294967295 x i32] }>, ptr addrspace(1) null, i64 0, i32 0, i64 %i4
+  %i6 = load i32, ptr addrspace(1) %i5, align 4
+  %i7 = add nuw i32 %arg, 2
+  %i8 = zext i32 %i7 to i64
+  %i9 = getelementptr <{ [4294967295 x i32] }>, ptr addrspace(1) null, i64 0, i32 0, i64 %i8
+  %i10 = load i32, ptr addrspace(1) %i9, align 4
+  %i11 = or i32 %i2, %i6
+  %i12 = or i32 %i10, %i11
+  store i32 %i12, ptr addrspace(1) null, align 4
+  ret void
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/load-constant-always-uniform.ll:62]: Can you simplify this test? I would expect this IR to already be vectorized

File: llvm/test/CodeGen/AMDGPU/load-constant-always-uniform.ll [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+; GFX12-NEXT:    v_mov_b32_e32 v2, s0
+; GFX12-NEXT:    s_delay_alu instid0(VALU_DEP_1)
+; GFX12-NEXT:    v_or3_b32 v2, v2, s1, s2
+; GFX12-NEXT:    global_store_b32 v[0:1], v2, off
+; GFX12-NEXT:    s_nop 0
+; GFX12-NEXT:    s_sendmsg sendmsg(MSG_DEALLOC_VGPRS)
+; GFX12-NEXT:    s_endpgm
+bb:
+  %i = zext i32 %arg to i64
+  %i1 = getelementptr i32, ptr addrspace(1) null, i64 %i
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/load-constant-always-uniform.ll:49]: non-null base pointer would be better

**Review by arsenm — APPROVED** [review]

**Pull request merged**